pkg/csi/service/server.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	csitypes "sigs.k8s.io/vsphere-csi-driver/pkg/csi/types"
+)
+
+var (
+	stopOnce sync.Once
+)
+
+// NonBlockingGRPCServer defines non-blocking GRPC server interfaces.
+type NonBlockingGRPCServer interface {
+	// Start services at the endpoint
+	Start(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer)
+
+	// Stop stops the gRPC server. It immediately closes all open
+	// connections and listeners.
+	// It cancels all active RPCs on the server side and the corresponding
+	// pending RPCs on the client side will get notified by connection
+	// errors.
+	Stop()
+
+	// GracefulStop stops the gRPC server gracefully. It stops the server
+	// from accepting new connections and RPCs and blocks until all the
+	// pending RPCs are finished.
+	GracefulStop()
+}
+
+// NewNonBlockingGRPCServer returns an instance of nonBlockingGRPCServer.
+func NewNonBlockingGRPCServer() NonBlockingGRPCServer {
+	return &nonBlockingGRPCServer{}
+}
+
+// nonBlockingGRPCServer implements the interface NonBlockingGRPCServer.
+type nonBlockingGRPCServer struct {
+	server *grpc.Server
+}
+
+func (s *nonBlockingGRPCServer) Start(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer) {
+	log := logger.GetLoggerWithNoContext()
+	if err := s.serve(endpoint, ids, cs, ns); err != nil {
+		log.Errorf("failed to start grpc server. Err: %v", err)
+	}
+}
+
+func (s *nonBlockingGRPCServer) GracefulStop() {
+	log := logger.GetLoggerWithNoContext()
+	stopOnce.Do(func() {
+		if s.server != nil {
+			s.server.Stop()
+		}
+		log.Info("gracefully stopped")
+	})
+}
+
+func (s *nonBlockingGRPCServer) Stop() {
```

Comment:
Whats the difference here with `GracefulStop`?

Reply:
`GracefulStop` was meant to call grpc's GracefulStop() method. Thanks for pointing out, corrected now.